internal/controllers/tlsedge_controller.go

```diff
@@ -0,0 +1,374 @@
+	"github.com/ngrok/ngrok-api-go/v5"
+)
+
+// TLSEdgeReconciler reconciles a TLSEdge object
+type TLSEdgeReconciler struct {
+	client.Client
+
+	Log      logr.Logger
+	Scheme   *runtime.Scheme
+	Recorder record.EventRecorder
+
+	ipPolicyResolver
+
+	NgrokClientset ngrokapi.Clientset
+
+	controller *baseController[*ingressv1alpha1.TLSEdge]
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *TLSEdgeReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	r.ipPolicyResolver = ipPolicyResolver{client: mgr.GetClient()}
+
+	r.controller = &baseController[*ingressv1alpha1.TLSEdge]{
+		Kube:     r.Client,
+		Log:      r.Log,
+		Recorder: r.Recorder,
+
+		kubeType: "v1alpha1.TLSEdge",
+		statusID: func(cr *ingressv1alpha1.TLSEdge) string { return cr.Status.ID },
+		create:   r.create,
+		update:   r.update,
+		delete:   r.delete,
+	}
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&ingressv1alpha1.TLSEdge{}).
+		Watches(
+			&source.Kind{Type: &ingressv1alpha1.IPPolicy{}},
+			handler.EnqueueRequestsFromMapFunc(r.listTLSEdgesForIPPolicy),
+		).
+		Complete(r)
+}
+
+//+kubebuilder:rbac:groups=ingress.k8s.ngrok.com,resources=tlsedges,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=ingress.k8s.ngrok.com,resources=tlsedges/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=ingress.k8s.ngrok.com,resources=tlsedges/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.13.1/pkg/reconcile
+func (r *TLSEdgeReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	return r.controller.reconcile(ctx, req, new(ingressv1alpha1.TLSEdge))
+}
+
+func (r *TLSEdgeReconciler) create(ctx context.Context, edge *ingressv1alpha1.TLSEdge) error {
+	if err := r.reconcileTunnelGroupBackend(ctx, edge); err != nil {
+		return err
+	}
+
+	// Try to find the edge by the backend labels
+	resp, err := r.findEdgeByBackendLabels(ctx, edge.Spec.Backend.Labels)
+	if err != nil {
+		return err
+	}
+
+	if resp != nil {
+		return r.updateEdgeStatus(ctx, edge, resp)
+	}
+
+	// No edge has been created for this edge, create one
+	r.Log.Info("Creating new TLSEdge", "namespace", edge.Namespace, "name", edge.Name)
+	resp, err = r.NgrokClientset.TLSEdges().Create(ctx, &ngrok.TLSEdgeCreate{
+		Hostports:   edge.Spec.Hostports,
+		Description: edge.Spec.Description,
+		Metadata:    edge.Spec.Metadata,
+		Backend: &ngrok.EndpointBackendMutate{
+			BackendID: edge.Status.Backend.ID,
+		},
+	})
+	if err != nil {
+		return err
+	}
+	r.Log.Info("Created new TLSEdge", "edge.ID", resp.ID, "name", edge.Name, "namespace", edge.Namespace)
+
+	if err := r.updateEdgeStatus(ctx, edge, resp); err != nil {
+		return err
+	}
+
+	return r.updateIPRestrictionRouteModule(ctx, edge, resp)
```

Comment:
are we missing `setTLSTermination` and `setMutualTLS` here?

Reply:
Yeah, and I think we're missing `updateIPRestriction` in `update`?